File: VeterinariaLosDogos/app/src/main/java/com/example/myapplication/RegisterPetActivity.java
package com.example.myapplication;

import android.app.DatePickerDialog;
import android.app.ProgressDialog;
import android.os.Bundle;
import android.text.TextUtils;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.myapplication.R;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.example.myapplication.models.Pet;


import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

public class RegisterPetActivity extends AppCompatActivity {

    private EditText etPetName, etBreed, etBirthdate;
    private Spinner spSpecies;
    private Button btnSavePet;
    private ProgressDialog progress;
    private DatabaseReference db;
    private FirebaseAuth mAuth;

    private final Calendar calendar = Calendar.getInstance();
    private final SimpleDateFormat isoFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register_pet); // crea este layout

        etPetName = findViewById(R.id.etPetName);
        etBreed = findViewById(R.id.etBreed);
        etBirthdate = findViewById(R.id.etBirthdate);
        spSpecies = findViewById(R.id.spSpecies);
        btnSavePet = findViewById(R.id.btnSavePet);

        // Spinner species
        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(this,
                R.array.species_array, android.R.layout.simple_spinner_item); // define array en strings.xml
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spSpecies.setAdapter(adapter);

        progress = new ProgressDialog(this);
        progress.setMessage("Guardando mascota...");

        db = FirebaseDatabase.getInstance().getReference();
        mAuth = FirebaseAuth.getInstance();

        etBirthdate.setOnClickListener(v -> showDatePicker());

        btnSavePet.setOnClickListener(v -> savePet());
    }

    private void showDatePicker() {
        int y = calendar.get(Calendar.YEAR);
        int m = calendar.get(Calendar.MONTH);
        int d = calendar.get(Calendar.DAY_OF_MONTH);
        DatePickerDialog dp = new DatePickerDialog(this, (view, year, month, dayOfMonth) -> {
            calendar.set(Calendar.YEAR, year);
            calendar.set(Calendar.MONTH, month);
            calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
            etBirthdate.setText(isoFormat.format(calendar.getTime()));
        }, y, m, d);
        dp.show();
    }

    private void savePet() {
        String petName = etPetName.getText().toString().trim();
        String breed = etBreed.getText().toString().trim();
        String species = spSpecies.getSelectedItem() != null ? spSpecies.getSelectedItem().toString() : "Otro";
        String birthdate = etBirthdate.getText().toString().trim();

        if (TextUtils.isEmpty(petName)) {
            etPetName.setError("Nombre de la mascota requerido");
            etPetName.requestFocus();
            return;
        }
        if (TextUtils.isEmpty(birthdate)) {
            etBirthdate.setError("Fecha de nacimiento requerida");
            etBirthdate.requestFocus();
            return;
        }

        String uid = mAuth.getCurrentUser() != null ? mAuth.getCurrentUser().getUid() : null;
        if (uid == null) {
            Toast.makeText(this, "Usuario no autenticado", Toast.LENGTH_SHORT).show();
            return;
        }

        progress.show();
        String petId = db.child("pets").push().getKey();
        Pet pet = new Pet(petId, uid, petName, species, breed, birthdate);
        db.child("pets").child(petId).setValue(pet)
                .addOnCompleteListener(task -> {
                    progress.dismiss();
                    if (task.isSuccessful()) {
                        Toast.makeText(RegisterPetActivity.this, "Mascota guardada", Toast.LENGTH_SHORT).show();
                        finish(); // vuelve al perfil/main
                    } else {
                        Toast.makeText(RegisterPetActivity.this, "Error guardando mascota: " + task.getException().getMessage(), Toast.LENGTH_LONG).show();
                    }
                });
    }
}

File: VeterinariaLosDogos/app/src/main/java/com/example/myapplication/RegisterPetActivity.kt
package com.example.myapplication.fragments

import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.provider.MediaStore
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.Fragment
import com.example.myapplication.R
import com.example.myapplication.models.Pet
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.storage.FirebaseDatabase
import com.google.firebase.storage.FirebaseStorage
import java.io.ByteArrayOutputStream
import java.util.*

class RegisterPetFragment : Fragment() {
    private lateinit var imgPet: ImageView
    private var imageBitmap: Bitmap? = null
    private val storage = FirebaseStorage.getInstance().reference
    private val db = FirebaseDatabase.getInstance().reference
    private val auth = FirebaseAuth.getInstance()

    // Lanzador para la cámara
    private val cameraLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            val bitmap = result.data?.extras?.get("data") as Bitmap
            imgPet.setImageBitmap(bitmap)
            imageBitmap = bitmap
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.fragment_register_pet, container, false)

        imgPet = v.findViewById(R.id.imgPetPreview) // Debes agregar este ImageView a tu XML
        val btnCamera = v.findViewById<Button>(R.id.btnCamera)
        val btnSave = v.findViewById<Button>(R.id.btnSavePet)

        btnCamera.setOnClickListener {
            val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            cameraLauncher.launch(takePictureIntent)
        }

        btnSave.setOnClickListener { uploadImageAndSavePet(v) }

        return v
    }

    private fun uploadImageAndSavePet(v: View) {
        val name = v.findViewById<EditText>(R.id.etPetName).text.toString()
        val uid = auth.currentUser?.uid ?: return
        val petId = db.child("pets").push().key ?: return

        if (imageBitmap != null) {
            val baos = ByteArrayOutputStream()
            imageBitmap!!.compress(Bitmap.CompressFormat.JPEG, 100, baos)
            val data = baos.toByteArray()

            // Subir a Firebase Storage
            val fileRef = storage.child("pets/$petId.jpg")
            fileRef.putBytes(data).addOnSuccessListener {
                fileRef.downloadUrl.addOnSuccessListener { uri ->
                    saveToDatabase(petId, uid, name, uri.toString())
                }
            }
        } else {
            saveToDatabase(petId, uid, name, "")
        }
    }

    private fun saveToDatabase(id: String, owner: String, name: String, url: String) {
        val pet = Pet(id, owner, name, imageUrl = url)
        db.child("pets").child(id).setValue(pet).addOnSuccessListener {
            Toast.makeText(context, "Mascota guardada con foto", Toast.LENGTH_SHORT).show()
        }
    }
}